pageserver/src/tenant.rs

```diff
@@ -1570,26 +1575,49 @@ impl Tenant {
             .enter()
             .map_err(|_| CreateTimelineError::ShuttingDown)?;
 
-        if let Ok(existing) = self.get_timeline(new_timeline_id, false) {
-            debug!("timeline {new_timeline_id} already exists");
-
-            if let Some(remote_client) = existing.remote_client.as_ref() {
-                // Wait for uploads to complete, so that when we return Ok, the timeline
-                // is known to be durable on remote storage. Just like we do at the end of
-                // this function, after we have created the timeline ourselves.
-                //
-                // We only really care that the initial version of `index_part.json` has
-                // been uploaded. That's enough to remember that the timeline
-                // exists. However, there is no function to wait specifically for that so
-                // we just wait for all in-progress uploads to finish.
-                remote_client
-                    .wait_completion()
-                    .await
-                    .context("wait for timeline uploads to complete")?;
+        // Get exclusive access to the timeline ID: this ensures that it does not already exist,
+        // and that no other creation attempts will be allowed in while we are working.  The
+        // uninit_mark is a guard.
+        let uninit_mark = match self.create_timeline_uninit_mark(new_timeline_id) {
+            Ok(m) => m,
+            Err(TimelineExclusionError::AlreadyCreating) => {
+                // Creation is in progress, we cannot create it again, and we cannot
+                // check if this request matches the existing one, so caller must try
+                // again later.
+                return Err(CreateTimelineError::AlreadyCreating);
+            }
+            Err(TimelineExclusionError::Other(e)) => {
+                return Err(CreateTimelineError::Other(e));
             }
+            Err(TimelineExclusionError::AlreadyExists(existing)) => {
+                debug!("timeline {new_timeline_id} already exists");
 
-            return Err(CreateTimelineError::AlreadyExists);
-        }
+                // Idempotency: creating the same timeline twice is not an error, unless
+                // the second creation has different parameters.
+                if existing.get_ancestor_timeline_id() != ancestor_timeline_id
+                    || existing.pg_version != pg_version
+                {
+                    return Err(CreateTimelineError::Conflict);
+                }
```

Comment:
`ancestor_start_lsn` should probably be included here?

I also wonder if this check could be under the wait_completion()? I did kind of like the upload waiting firing off always. It would risk another timeout, but at least we would be pacing the errors?

Reply:
> ancestor_start_lsn should probably be included here?

e87ffc0d1

> I also wonder if this check could be under the wait_completion()? I did kind of like the upload waiting firing off always. It would risk another timeout, but at least we would be pacing the errors?

I think this case should be very rare: the control plane will never do it because it is always retrying with the same parameters.  If we waited for the existing timeline to upload, that would delay the retry one time while the timeline is still ccreating, but then if a caller keeps retrying they'll proceed immediately next time around.

Reply:
True, it wouldn't help more than the first uploads. Resolving.